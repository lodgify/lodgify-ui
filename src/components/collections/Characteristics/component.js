import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'semantic-ui-react';

import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { getFirstFourItems } from 'utils/get-first-four-items';
import { Grid } from 'layout/Grid';
import { GridColumn } from 'layout/GridColumn';
import { Icon } from 'elements/Icon';
import { ShowOn } from 'layout/ShowOn';

/**
 * The standard widget for displaying a collection of characteristics for a property or a room.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({ spaceCharacteristics }) => (
  <Grid columns={1}>
    <GridColumn width={12}>
      <ShowOn
        computer
        parent={List}
        parentProps={{ horizontal: true }}
        tablet
        widescreen
      >
        {getFirstFourItems(spaceCharacteristics).map(
          ({ iconName, text }, index) => (
            <List.Item key={buildKeyFromStrings(text, index)}>
              <Icon labelText={text} name={iconName} />
            </List.Item>
          )
        )}
      </ShowOn>
      <ShowOn mobile parent={Grid} parentProps={{ columns: 1 }}>
        {getFirstFourItems(spaceCharacteristics).map(
          ({ iconName, text }, index) => (
            <GridColumn
              computer={3}
              key={buildKeyFromStrings(text, index)}
              mobile={6}
            >
              <Icon labelText={text} name={iconName} />
            </GridColumn>
          )
        )}
      </ShowOn>
    </GridColumn>
  </Grid>
);

Component.displayName = 'Characteristics';

Component.defaultProps = {
  spaceCharacteristics: [],
};

Component.propTypes = {
  /** The main characteristics to display. Maximum four. */
  spaceCharacteristics: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The name of the icon to display.
       * [See here for the full list of valid icon names](https://github.com/lodgify/lodgify-ui/blob/master/src/components/elements/Icon/constants.js)
       */
      iconName: PropTypes.string.isRequired,
      /** A visible label to display for the key fact. */
      text: PropTypes.string.isRequired,
    })
  ),
};
