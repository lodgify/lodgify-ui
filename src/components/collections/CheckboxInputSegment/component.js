import React from 'react';
import PropTypes from 'prop-types';
import { Segment } from 'semantic-ui-react';

import { Divider } from 'elements/Divider';
import { FlexContainer } from 'layout/FlexContainer';
import { Grid } from 'layout/Grid';
import { Heading } from 'typography/Heading';
import { Checkbox } from 'inputs/Checkbox';
import { GridColumn } from 'layout/GridColumn';
import { withResponsive } from 'utils/with-responsive';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';

import { getGridColumns } from './utils/getGridColumns';

/**
 * A collection of information alongside checkboxes.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
const Component = ({ heading, checkboxes, isUserOnMobile }) => (
  <Segment className="is-input-segment is-checkbox-input-segment" vertical>
    <FlexContainer
      alignItems="baseline"
      flexDirection="column"
      justifyContent="space-between"
    >
      <Heading>{heading}</Heading>
      <Divider size="small" />
      <Grid columns={getGridColumns(isUserOnMobile)}>
        {checkboxes.map(({ name, label, ...checkboxProps }) => (
          <GridColumn key={buildKeyFromStrings(label, name)}>
            <Checkbox
              label={label}
              name={name}
              {...checkboxProps}
              isFluid={isUserOnMobile}
              isLabelLeft={isUserOnMobile}
            />
          </GridColumn>
        ))}
      </Grid>
    </FlexContainer>
  </Segment>
);

Component.displayName = 'CheckboxInputSegment';

Component.propTypes = {
  /** The checkboxes to display. See [the `Checkbox` component for valid props](#/Inputs/Checkbox).  */
  checkboxes: PropTypes.arrayOf(PropTypes.object).isRequired,
  /** The text for the input segment's heading. */
  heading: PropTypes.string.isRequired,
  /**
   * Is the user on a mobile device.
   * Provided by `withResponsive` so ignored in the styleguide.
   * @ignore
   */
  isUserOnMobile: PropTypes.bool.isRequired,
};

export const ComponentWithResponsive = withResponsive(Component);
