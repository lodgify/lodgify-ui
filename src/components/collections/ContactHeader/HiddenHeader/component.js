import React, { useRef, useState, useEffect } from 'react';
import {
  string,
  arrayOf,
  shape,
  func,
  bool,
  oneOfType,
  number,
} from 'prop-types';
import { Menu, Accordion } from 'semantic-ui-react';
import { size } from 'lodash';

import { Heading } from 'typography/Heading';
import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { Modal } from 'elements/Modal';
import { Submenu } from 'elements/Submenu';
import { testidFactory } from 'utils/testid';

import { PHONE_PREFIX } from '../constants';

const TEST_ID_PREFIX = 'hiddenHeader';

const testid = testidFactory(TEST_ID_PREFIX);

export const Component = ({
  headingHref,
  headingText,
  navigationItems,
  languageOptions,
  onChangeLanguage,
  languageValue,
  currencyOptions,
  currencyNoResultsText,
  onChangeCurrency,
  currencyValue,
  phoneNumber,
}) => {
  const [activeItem, setActiveItem] = useState('');

  useEffect(() => {
    /* istanbul ignore next */
    if (navigationItems.length > 0) {
      setActiveItem(navigationItems[0].text);
    }
  }, [navigationItems.length]);

  const menu = useRef();

  const onMenuLinkClick = text => {
    setActiveItem(text);
  };

  return (
    <Modal
      className="hidden-contact-header-modal"
      closeIcon={<Icon labelText="Close" name="close" />}
      data-testid={testid()}
      header={
        <Menu.Item href={headingHref} link>
          <Heading className="header-title" size="small">
            {headingText}
          </Heading>
        </Menu.Item>
      }
      isFullscreen
      trigger={<Icon name={ICON_NAMES.BARS} />}
    >
      <div ref={menu}>
        <Menu text vertical>
          {navigationItems.map(({ subItems, text, link }, index) =>
            size(subItems) ? (
              <Accordion
                as={Menu.Item}
                key={buildKeyFromStrings(text, index)}
                panels={[
                  {
                    title: {
                      content: text,
                      key: buildKeyFromStrings(text, index),
                    },
                    content: {
                      content: subItems.map(({ text, href }, index) => (
                        <Menu.Item
                          active={text === activeItem}
                          data-testid={testid(`accordionItem_${index}`)}
                          href={href}
                          key={buildKeyFromStrings(text, index)}
                          link
                          onClick={() => onMenuLinkClick(text)}
                        >
                          {text}
                        </Menu.Item>
                      )),
                      key: buildKeyFromStrings(index, text),
                    },
                  },
                ]}
              />
            ) : (
              <Menu.Item
                active={text === activeItem}
                data-testid={testid(`hiddenHeaderItem_${index}`)}
                href={link}
                key={buildKeyFromStrings(text, index)}
                link
                onClick={() => onMenuLinkClick(text)}
              >
                {text}
              </Menu.Item>
            )
          )}
        </Menu>
        <Menu vertical>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
      </div>
    </Modal>
  );
};

Component.defaultProps = {
  currencyNoResultsText: '',
  currencyOptions: [],
  currencyValue: null,
  headingHref: '',
  headingText: '',
  languageOptions: [],
  languageValue: null,
  phoneNumber: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The link for the heading. */
  headingHref: string,
  /** The text for the heading. */
  headingText: string,
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The items for a user to navigate the site. */
  navigationItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
};
