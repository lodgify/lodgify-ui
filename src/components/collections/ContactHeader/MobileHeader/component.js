import React, { useState, useEffect } from 'react';
import {
  string,
  arrayOf,
  shape,
  func,
  bool,
  oneOfType,
  number,
  oneOf,
} from 'prop-types';
import { Menu, Accordion } from 'semantic-ui-react';
import { size } from 'lodash';

import { buildKeyFromStrings } from 'utils/build-key-from-strings';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { Modal } from 'elements/Modal';
import { Submenu } from 'elements/Submenu';
import { testidFactory } from 'utils/testid';
import { Logo } from 'elements/Logo';

import { PHONE_PREFIX } from '../constants';

const TEST_ID_PREFIX = 'MobileHeader';

const testid = testidFactory(TEST_ID_PREFIX);

export const Component = ({
  currencyNoResultsText,
  currencyOptions,
  currencyValue,
  languageOptions,
  languageValue,
  logoHref,
  logoSize,
  logoSizes,
  logoSrc,
  logoSrcSet,
  logoSubText,
  logoText,
  navigationItems,
  onChangeCurrency,
  onChangeLanguage,
  phoneNumber,
}) => {
  const [activeItem, setActiveItem] = useState('');

  useEffect(() => {
    /* istanbul ignore next */
    if (navigationItems.length > 0) {
      setActiveItem(navigationItems[0].text);
    }
  }, [navigationItems.length]);

  const onMenuLinkClick = text => {
    setActiveItem(text);
  };

  return (
    <Modal
      className="hidden-contact-header-modal"
      closeIcon={<Icon labelText="Close" name="close" />}
      data-testid={testid()}
      header={
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
        />
      }
      isFullscreen
      trigger={<Icon name={ICON_NAMES.BARS} />}
    >
      <div>
        <Menu text vertical>
          {navigationItems.map(({ subItems, text, href }, index) =>
            size(subItems) ? (
              <Accordion
                as={Menu.Item}
                key={buildKeyFromStrings(text, index)}
                panels={[
                  {
                    title: {
                      content: text,
                      key: buildKeyFromStrings(text, index),
                    },
                    content: {
                      content: subItems.map(({ text, href }, index) => (
                        <Menu.Item
                          active={text === activeItem}
                          data-testid={testid(`accordionItem_${index}`)}
                          href={href}
                          key={buildKeyFromStrings(text, index)}
                          link
                          onClick={() => onMenuLinkClick(text)}
                        >
                          {text}
                        </Menu.Item>
                      )),
                      key: buildKeyFromStrings(index, text),
                    },
                  },
                ]}
              />
            ) : (
              <Menu.Item
                active={text === activeItem}
                data-testid={testid(`MobileHeaderItem_${index}`)}
                href={href}
                key={buildKeyFromStrings(text, index)}
                link
                onClick={() => onMenuLinkClick(text)}
              >
                {text}
              </Menu.Item>
            )
          )}
        </Menu>
        <Menu vertical>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
      </div>
    </Modal>
  );
};

Component.defaultProps = {
  currencyNoResultsText: '',
  currencyOptions: [],
  currencyValue: null,
  languageOptions: [],
  languageValue: null,
  logoHref: '',
  logoSize: 'medium',
  logoSizes: undefined,
  logoSrc: undefined,
  logoSrcSet: undefined,
  logoSubText: '',
  logoText: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
  phoneNumber: '',
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The href for the logo link. */
  logoHref: string,
  /** The maximum size of the headers logo. */
  logoSize: oneOf(['medium', 'large', 'huge']),
  /** A list of one or more strings separated by commas indicating a set of source sizes for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSizes: string,
  /** The src url for the logo. */
  logoSrc: string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSrcSet: string,
  /** The text that appears under the logo or logo text. */
  logoSubText: string,
  /** The text for the logo. */
  logoText: string,
  /** The items for a user to navigate the site. */
  /** The navigation items displayed in the header. */
  navigationItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
};
