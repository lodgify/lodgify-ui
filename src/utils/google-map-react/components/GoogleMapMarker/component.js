import React from 'react';
import PropTypes from 'prop-types';

export const Component = ({ imageSrc, style }) => (
  <img src={imageSrc} style={style} />
);

Component.displayName = 'GoogleMapMarker';

Component.propTypes = {
  /** A url pointing to the image to be used as a marker. */
  imageSrc: PropTypes.string.isRequired,
  /** Inline styles for the marker. */
  style: PropTypes.shape({
    transform: PropTypes.string.isRequired,
  }).isRequired,
};
