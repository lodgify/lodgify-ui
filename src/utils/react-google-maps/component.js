import React from 'react';
import PropTypes from 'prop-types';
import { GoogleMap } from 'react-google-maps/lib/components/GoogleMap';
import { Marker } from 'react-google-maps/lib/components/Marker';
import { Circle } from 'react-google-maps/lib/components/Circle';

import { mapOptions, circleOptions } from './constants';
import { adaptCoordinates } from './utils/adaptCoordinates';

/**
 * Canonical implementation for react-google-maps.
 * See https://tomchentw.github.io/react-google-maps/#usage--configuration
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  isShowingExactLocation,
  isShowingApproximateLocation,
  latitude,
  longitude,
}) => {
  return (
    <GoogleMap
      center={adaptCoordinates(latitude, longitude)}
      options={mapOptions}
    >
      {!!isShowingExactLocation && (
        <Marker position={adaptCoordinates(latitude, longitude)} />
      )}
      {!!isShowingApproximateLocation && (
        <Circle
          center={adaptCoordinates(latitude, longitude)}
          options={circleOptions}
        />
      )}
    </GoogleMap>
  );
};

Component.displayName = 'ReactGoogleMap';

Component.propTypes = {
  /** Is the map showing a marker for the approximate location. */
  isShowingApproximateLocation: PropTypes.bool.isRequired,
  /** Is the map showing a marker for the exact location. */
  isShowingExactLocation: PropTypes.bool.isRequired,
  /** The latitude coordinate for the center of the map and/or location of the marker */
  latitude: PropTypes.number.isRequired,
  /** The longitude coordinate for the center of the map and/or location of the marker */
  longitude: PropTypes.number.isRequired,
};
