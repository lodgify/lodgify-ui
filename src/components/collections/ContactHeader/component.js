import React from 'react';
import { Menu } from 'semantic-ui-react';
import {
  string,
  arrayOf,
  shape,
  oneOfType,
  oneOf,
  number,
  bool,
  func,
} from 'prop-types';

import { FlexContainer } from 'layout/FlexContainer';
import { Button } from 'elements/Button';
import { Submenu } from 'elements/Submenu';
import { size } from 'utils/size';
import { Icon, ICON_NAMES } from 'elements/Icon';
import { ShowOn } from 'layout/ShowOn';
import { testidFactory } from 'utils/testid';
import { Logo } from 'elements/Logo';

import { HorizontalMenu } from '../HorizontalMenu';

import { MobileHeader } from './MobileHeader';
import { PHONE_PREFIX } from './constants';

const TEST_ID_PREFIX = 'contactHeader';

const testid = testidFactory(TEST_ID_PREFIX);

/**
 * A header displays a logo, grouped navigation items
 * contact information and an a primary call to action.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  primaryButtonText,
  navigationMenuItems,
  languageOptions,
  onChangeLanguage,
  currencyOptions,
  onChangeCurrency,
  currencyValue,
  currencyNoResultsText,
  languageValue,
  phoneNumber,
  logoHref,
  logoSubText,
  logoSize,
  logoSrc,
  logoSizes,
  logoText,
  logoSrcSet,
}) => (
  <header className="contact-header" data-testid={testid()}>
    <ShowOn computer tablet widescreen>
      <FlexContainer alignItems="center" justifyContent="space-between">
        <Menu>
          {size(languageOptions) > 1 && (
            <Menu.Item>
              <Submenu
                items={languageOptions}
                name="language"
                onChange={onChangeLanguage}
                value={languageValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {size(currencyOptions) > 1 && (
            <Menu.Item>
              <Submenu
                isSearchable
                items={currencyOptions}
                name="currency"
                noResultsText={currencyNoResultsText}
                onChange={onChangeCurrency}
                value={currencyValue}
                willOpenAbove
              />
            </Menu.Item>
          )}
          {!!phoneNumber && (
            <Menu.Item className="is-selectable">
              <a href={`${PHONE_PREFIX}${phoneNumber.replace(/\s/g, '')}`}>
                <Icon labelText={phoneNumber} name={ICON_NAMES.PHONE} />
              </a>
            </Menu.Item>
          )}
        </Menu>
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoSubTextAlignment="center"
          logoText={logoText}
        />
        <Button>{primaryButtonText}</Button>
      </FlexContainer>
      <HorizontalMenu isHeader items={navigationMenuItems} />
    </ShowOn>
    <ShowOn mobile>
      <FlexContainer
        alignItems="center"
        className="hidden-contact-header"
        justifyContent="space-between"
      >
        <Logo
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
        />
        <MobileHeader
          currencyNoResultsText={currencyNoResultsText}
          currencyOptions={currencyOptions}
          currencyValue={currencyValue}
          languageOptions={languageOptions}
          languageValue={languageValue}
          logoHref={logoHref}
          logoSize={logoSize}
          logoSizes={logoSizes}
          logoSrc={logoSrc}
          logoSrcSet={logoSrcSet}
          logoSubText={logoSubText}
          logoText={logoText}
          navigationItems={navigationMenuItems}
          onChangeCurrency={onChangeCurrency}
          phoneNumber={phoneNumber}
        />
      </FlexContainer>
    </ShowOn>
  </header>
);

Component.displayName = 'ContactHeader';

Component.defaultProps = {
  currencyNoResultsText: undefined,
  currencyOptions: [],
  currencyValue: null,
  languageOptions: [],
  languageValue: null,
  logoHref: '',
  logoSize: 'medium',
  logoSizes: undefined,
  logoSrc: undefined,
  logoSrcSet: undefined,
  logoSubText: '',
  logoText: '',
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
  phoneNumber: '',
  primaryButtonText: '',
};

Component.propTypes = {
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The href for the logo link. */
  logoHref: string,
  /** The maximum size of the headers logo. */
  logoSize: oneOf(['medium', 'large', 'huge']),
  /** A list of one or more strings separated by commas indicating a set of source sizes for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSizes: string,
  /** The src url for the logo. */
  logoSrc: string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the logo. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  logoSrcSet: string,
  /** The text that appears under the logo or logo text. */
  logoSubText: string,
  /** The text for the logo. */
  logoText: string,
  /** The items for a user to navigate the site. */
  navigationMenuItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
  /** The text to display for the primary button. */
  primaryButtonText: string,
};
