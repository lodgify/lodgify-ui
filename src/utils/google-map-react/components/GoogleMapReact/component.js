import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'fast-deep-equal';
import GoogleMapReact from 'google-map-react';
import { fitBounds } from 'google-map-react/utils';
import debounce from 'debounce';

import {
  DEFAULT_ZOOM,
  MARKER_IMAGE_SRC,
  MARKER_IMAGE_STYLE,
  CIRCLE_IMAGE_SRC,
  CIRCLE_IMAGE_STYLE,
} from '../../constants';
import { getMapOptions } from '../../utils/getMapOptions';
import { GoogleMapMarker } from '../GoogleMapMarker';

import { adaptENSWtoNESW } from './utils/adaptENSWtoNESW';
import { getCenterFromBounds } from './utils/getCenterFromBounds';
import { adaptNESWtoENSW } from './utils/adaptNESWtoENSW';
import { getCenter } from './utils/getCenter';
import { getZoomedState } from './utils/getZoomedState';

export class Component extends PureComponent {
  state = {
    areBoundsChanged: false,
    areBoundsChangedProgramatically: false,
    bounds: null,
    center: null,
    isDragged: false,
    isZoomed: false,
    size: null,
    zoom: DEFAULT_ZOOM,
  };

  componentDidMount = () => {
    const { bounds } = this.props;

    if (!bounds) return;

    this.setState({
      bounds: adaptENSWtoNESW(bounds),
      center: getCenterFromBounds(bounds),
    });
  };

  componentDidUpdate = ({ bounds: previousControlledBounds }) => {
    const { areBoundsChanged, bounds, isDragged, isZoomed, size } = this.state;
    const { bounds: controlledBounds, onBoundsChange } = this.props;

    if (areBoundsChanged) {
      onBoundsChange(adaptNESWtoENSW(bounds), isZoomed || isDragged);
      this.setState({
        areBoundsChanged: false,
        areBoundsChangedProgramatically: false,
        isDragged: false,
        isZoomed: false,
      });
    }

    if (isEqual(previousControlledBounds, controlledBounds)) return;

    const { center, newBounds, zoom } = fitBounds(
      adaptENSWtoNESW(controlledBounds),
      size
    );

    this.setState({
      areBoundsChangedProgramatically: true,
      bounds: newBounds,
      center,
      zoom,
    });
  };

  handleChange = ({ bounds, center, size, zoom }) => {
    if (!this.props.bounds) return;
    this.setState({ areBoundsChanged: true, bounds, center, size, zoom });
  };

  handleDrag = debounce(() => {
    this.setState({ isDragged: true });
  });

  handleZoomAnimationEnd = () => this.setState(getZoomedState);

  render = () => {
    const {
      apiKey,
      hasDefaultStyles,
      isShowingExactLocation,
      isShowingApproximateLocation,
      latitude,
      longitude,
      markers,
    } = this.props;
    const { bounds, zoom } = this.state;
    const center = getCenter(this.state.center, latitude, longitude);

    if (!center) return null;

    return (
      <GoogleMapReact
        bootstrapURLKeys={{
          key: apiKey,
        }}
        bounds={bounds}
        center={center}
        onChange={this.handleChange}
        onDrag={this.handleDrag}
        onZoomAnimationEnd={this.handleZoomAnimationEnd}
        options={getMapOptions(hasDefaultStyles)}
        ref={this.createRef}
        zoom={zoom}
      >
        {!!isShowingExactLocation && (
          <GoogleMapMarker
            imageSrc={MARKER_IMAGE_SRC}
            lat={latitude}
            lng={longitude}
            style={MARKER_IMAGE_STYLE}
          />
        )}
        {!!isShowingApproximateLocation && (
          <GoogleMapMarker
            imageSrc={CIRCLE_IMAGE_SRC}
            lat={latitude}
            lng={longitude}
            style={CIRCLE_IMAGE_STYLE}
          />
        )}
        {markers.map(({ component, key, latitude, longitude }) =>
          React.cloneElement(component, { key, lat: latitude, lng: longitude })
        )}
      </GoogleMapReact>
    );
  };
}

Component.displayName = 'GoogleMapReact';

Component.defaultProps = {
  bounds: null,
  latitude: null,
  longitude: null,
};

Component.propTypes = {
  /** An [API key](https://developers.google.com/maps/documentation/javascript/get-api-key) for using Google Maps. */
  apiKey: PropTypes.string.isRequired,
  /** The bounds of the map as latitude and longitude values. */
  bounds: PropTypes.shape({
    east: PropTypes.number.isRequired,
    north: PropTypes.number.isRequired,
    south: PropTypes.number.isRequired,
    west: PropTypes.number.isRequired,
  }),
  /** Does the map have the default Google Maps styles. */
  hasDefaultStyles: PropTypes.bool.isRequired,
  /** Is the map showing a marker for the approximate location at the center of the map. */
  isShowingApproximateLocation: PropTypes.bool.isRequired,
  /** Is the map showing a marker for the exact location location at the center of the map. */
  isShowingExactLocation: PropTypes.bool.isRequired,
  /** The latitude coordinate for the center of the map. */
  latitude: PropTypes.number,
  /** The longitude coordinate for the center of the map. */
  longitude: PropTypes.number,
  /** The markers to display on the map. */
  markers: PropTypes.arrayOf(
    PropTypes.shape({
      /** The React component to show as a marker. */
      component: PropTypes.node.isRequired,
      /** A unique identifier for the marker component. */
      key: PropTypes.string.isRequired,
      /** The latitude coordinate for the marker. */
      latitude: PropTypes.number,
      /** The longitude coordinate for the marker. */
      longitude: PropTypes.number,
    })
  ).isRequired,
  /**
   * A function called when the bounds of a dynamic map change.
   * @param {Object}  bounds
   * @param {number}  bounds.east
   * @param {number}  bounds.north
   * @param {number}  bounds.south
   * @param {number}  bounds.west
   * @param {boolean} isUserAction
   */
  onBoundsChange: PropTypes.func.isRequired,
};
