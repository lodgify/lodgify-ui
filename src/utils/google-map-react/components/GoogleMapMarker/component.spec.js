import React from 'react';
import { shallow } from 'enzyme';

import { Component as GoogleMapMarker } from './component';

describe('GoogleMapMarker', () => {
  it('should render the right structure', () => {
    const wrapper = shallow(
      <GoogleMapMarker
        imageSrc="some imageSrc"
        style={{ transform: 'some transform' }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
