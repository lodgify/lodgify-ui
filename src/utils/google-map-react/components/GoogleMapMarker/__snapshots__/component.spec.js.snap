// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMapMarker should render the right structure 1`] = `
<img
  src="some imageSrc"
  style={
    Object {
      "transform": "some transform",
    }
  }
/>
`;
